### ISSUE TITLE
TypeScript class prototype includes instance member

### ISSUE THREAD

**Issue by luokuning** [Microsoft/TypeScript#30516]
**TypeScript Version:**  TypeScript official playground

**Search Terms:**

**Code**

```ts
class Hello {
    abc = {
        a: 'l',
        b: 1,
    }
    k = (prop: keyof typeof Hello.prototype.abc) => {
        console.log(prop)
    }
}
```

**Expected behavior:**

TypeScript compiler should complain about the `Hello.prototype.abc`, because `abc` is an instance property not in the `Hello.prototype`.

**Actual behavior:**

The code works fine in TypeScript playground, no compiling errors.

**Playground Link:** http://www.typescriptlang.org/play/index.html#src=class%20Hello%20%7B%0D%0A%20%20%20%20abc%20%3D%20%7B%0D%0A%20%20%20%20%20%20%20%20a%3A%20'l'%2C%0D%0A%20%20%20%20%20%20%20%20b%3A%201%2C%0D%0A%20%20%20%20%7D%0D%0A%20%20%20%20k%20%3D%20(prop%3A%20keyof%20typeof%20Hello.prototype.abc)%20%3D%3E%20%7B%0D%0A%20%20%20%20%20%20%20%20console.log(prop)%0D%0A%20%20%20%20%7D%0D%0A%7D

**Comment by RyanCavanaugh** [comment]
Duplicate #11558

**Comment by RyanCavanaugh** [comment]
Also #20922

**Comment by jcalz** [comment]
(I'm commenting here because the linked issues are locked)

One could imagine that instance-only properties could be made optional in the prototype type.  But I'm sure this would break all kinds of things (I'm pretty sure `x instanceof C` narrows `x` to `C['prototype']`, which would be a funny thing to do if the prototype were wider than `C`) and not add much value.
